libsolutil/LazyInit.h

```diff
@@ -0,0 +1,107 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolutil/Assertions.h>
+#include <libsolutil/Exceptions.h>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace solidity::util
+{
+
+DEV_SIMPLE_EXCEPTION(BadLazyInitAccess);
+
+/**
+ * Represents a lazy-initialized value.
+ * @tparam T the type of the lazy-initialized object; may not be a function, reference, array, or void type; may be const-qualified.
+ */
+template<typename T>
+class LazyInit
+{
+public:
+	using value_type = T;
+
+	static_assert(std::is_object_v<value_type>, "Function, reference, and void types are not supported");
+	static_assert(!std::is_array_v<value_type>, "Array types are not supported.");
+	static_assert(!std::is_volatile_v<value_type>, "Volatile-qualified types are not supported.");
+
+	LazyInit() = default;
+
+	// Copy constructor is okay because they will copy the optional's value.
```

Comment:
Do we need this to be copyable?

Reply:
No, it looks like everything still compiles with a deleted copy constructor.

I made it copyable because it makes philosophical sense for it to be copyable - it's a value that happens to be computed on first use, but it's still a value. I'll change it if you would prefer, though.